addons/test_website/tests/test_session.py

```diff
@@ -1,9 +1,39 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+from lxml import html
+
 import odoo.tests
-from odoo.tools import mute_logger
+
+from odoo.addons.base.tests.common import HttpCaseWithUserDemo
 
 
 @odoo.tests.common.tagged('post_install', '-at_install')
-class TestWebsiteSession(odoo.tests.HttpCase):
+class TestWebsiteSession(HttpCaseWithUserDemo):
 
     def test_01_run_test(self):
         self.start_tour('/', 'test_json_auth')
+
+    def test_branding_cache(self):
+        def has_branding(html_text):
+            el = html.fromstring(html_text)
+            return el.xpath('//*[@data-oe-model="test.model"]')
+
+        self.user_demo.groups_id += self.env.ref('website.group_website_restricted_editor')
+        self.user_demo.groups_id -= self.env.ref('website.group_website_designer')
+
+        # Create session for demo user.
+        public_session = self.authenticate(None, None)
+        demo_session = self.authenticate('demo', 'demo')
+        record = self.env['test.model'].search([])
+        result = self.url_open(f'/test_website/model_item/{record.id}')
+        self.assertTrue(has_branding(result.text), "Should have branding for user demo")
+
+        # Public user.
+        self.opener.cookies['session_id'] = public_session.sid
```

Comment:
The solution with different host (localhost and 127.0.0.1) was not working?
I liked it since it was not relying on our session_id implementation or any details of implementation at all and was simulating real use cases but it's a detail, just curious

Reply:
Indeed, it did not work: the `self.opener` kept using the same session id for calls on all hosts (localhost, 127.0.0.1...) So I ended up having to manipulate it manually unfortunately.